splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/BroadcastJoinOperation.java

```diff
@@ -178,7 +178,7 @@ public SpliceOperation getLeftOperation(){
          * so route to the rdd implementation for now for SSQ.  Also, spark join can't handle a zero length
          * hash key, so it will use rdd as well.
          */
-        if (rightFromSSQ || leftHashKeys.length == 0)
+        if (rightFromSSQ || (leftHashKeys.length == 0 && !hasSparkJoinPredicate()))
```

Comment:
Do we need to check for `varcharDB2CompatibilityMode` here? If it's enabled we won't apply the sparkJoinPredicate to the native join (it'll be applied as a non-native flat-map function after the full join is generated). Is that OK or not? I guess that's still what we'd be doing if we set `useDataset = true`, so maybe this is fine.

Reply:
@dgomezferro This is a great point.  I have added code to transform the sparkJoinPredicate into an column expression with rtrimmed columns, so varcharDB2CompatibilityMode can use full native spark too.

Reply:
Also, I have reworked the code in HashableJoinStrategy to allow broadcast join to be picked in more cases (which triggers BroadCastNestedLoopJoin in Spark).  Previously the rules were added because BroadcastJoinStrategy.feasible was called even for single-table access, so we ended up pick broadcast join for things which weren't joins.  
The new check:

> if (!isSingleTableScan(optimizer))

should avoid those cases without all the convoluted rules.